### PR TITLE
Add proxy support for ImageRepository API

### LINKED ISSUE #612: Add proxy support in ImageRepository
The ImageRepository API should provide a way to configure the proxy that should be used when connecting to an image registry. The ImageRepository reconciler should use the provided proxy configuration to connect to the remote registry server per object.

The API would be similar to the proxy secret reference in GitRepository API, refer https://fluxcd.io/flux/components/source/gitrepositories/#proxy-secret-reference.

Refer https://github.com/fluxcd/source-controller/pull/1536 for a reference to how it's implemented in OCIRepository which uses the same registry client.

### PR DESCRIPTION
**Pull request by matheuscscp**
Fixes #612

### CODE REVIEW

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @matheuscscp 🏅

File: internal/controller/imagerepository_controller_test.go [hunk 1]
```diff
@@ -498,19 +532,22 @@ func TestImageRepositoryReconciler_scan(t *testing.T) {
...
+
 	tests := []struct {
 		name           string
 		tags           []string
 		exclusionList  []string
 		annotation     string
 		db             *mockDatabase
-		wantErr        bool
+		proxyURL       *url.URL
+		err            string
```
[the remark this answers is not in the document]
Comment (matheuscscp) [reply internal/controller/imagerepository_controller_test.go]: Fixed, thanks!

File: internal/controller/imagerepository_controller_test.go [hunk 2]
```diff
@@ -607,6 +671,191 @@ func TestImageRepositoryReconciler_scan(t *testing.T) {
 	}
 }
 
+func TestImageRepositoryReconciler_getProxyURL(t *testing.T) {
+	tests := []struct {
+		name        string
+		repo        *imagev1.ImageRepository
+		objects     []client.Object
+		expectedURL string
+		expectedErr string
```
[the remark this answers is not in the document]
Comment (matheuscscp) [reply internal/controller/imagerepository_controller_test.go]: Fixed, thanks!

File: internal/controller/imagerepository_controller.go [hunk 3]
```diff
@@ -371,6 +373,8 @@ func (r *ImageRepositoryReconciler) setAuthOptions(ctx context.Context, obj *ima
 		options = append(options, remote.WithAuth(auth))
 	}
 
+	var transportOptions []func(*http.Transport)
```
[the remark this answers is not in the document]
Comment (matheuscscp) [reply internal/controller/imagerepository_controller.go]: PR updated with `github.com/fluxcd/pkg/oci v0.41.0`, still thinking about how to add proxy tests

File: internal/controller/imagerepository_controller.go [hunk 4]
```diff
@@ -357,7 +376,12 @@ func (r *ImageRepositoryReconciler) setAuthOptions(ctx context.Context, obj *ima
 		default:
 			opts = r.DeprecatedLoginOpts
 		}
-		auth, authErr = login.NewManager().Login(ctx, obj.Spec.Image, ref, opts)
+		var managerOpts []login.Option
+		if proxyURL != nil {
+			managerOpts = append(managerOpts, login.WithProxyURL(proxyURL))
+		}
```
Comment (stefanprodan) [inline internal/controller/imagerepository_controller.go:382]: Why not `login.NewManager(login.WithProxyURL(proxyURL)).Login(...)`, the nil check is done by the manager. This is how I did it in source-controller here: https://github.com/fluxcd/source-controller/pull/1607

Comment (matheuscscp) [reply internal/controller/imagerepository_controller.go:382]: I think `login.NewManager(login.WithProxyURL(proxyURL)).Login(...)` kinda suggests to the reader that there is always a proxy URL set in this case (and is also a long expression which would read nicer if broken down, but that's unrelated to your specific suggestion, we could just break it in `managerOpts := ...; manager := ...; manager.Login(...)` and still do it as you say). This is how I did it for most of these optional parameters for Bucket: https://github.com/fluxcd/source-controller/blob/main/internal/controller/bucket_controller.go#L449-L537

File: internal/controller/imagerepository_controller.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (matheuscscp) [reply internal/controller/imagerepository_controller.go]: As discussed in the dev meeting today, we will add e2e tests for OCI auth using proxy, but we won't block this PR on it.

**Pull request merged**